common/shared.c

```diff
@@ -234,6 +240,72 @@ uint64_t ft_info_to_mr_access(struct fi_info *info)
 	return mr_access;
 }
 
+#define bit_isset(x, i) (x >> i & 1)
+#define for_each_bit(x, i) for (i = 0; i < (8 * sizeof(x)); i++)
+
+static inline int bit_set_count(uint64_t val)
+{
+	int cnt = 0;
+	while (val) {
+		cnt++;
+		val &= val - 1;
+	}
+	return cnt;
+}
+
+static inline int contains_pattern(uint64_t *set, int len, uint64_t pat) {
+	int i;
+	for (i = 0; i < len; i++)
+		if (set[i] == pat) return 1;
+
+	return 0;
+}
+
+void ft_alloc_bit_combo(uint64_t fixed, uint64_t opt, uint64_t **combos)
+{
+	uint64_t *flags;
+	uint64_t toAdd;
+	int num_flags;
+	int num_comb;
+	int i, j, k, start, end;
+
+	num_flags = bit_set_count(opt);
+	flags = calloc(num_flags, sizeof(fixed));
+	*combos = calloc(1 << num_flags, sizeof(fixed));
+
+	num_flags = num_comb = 0;
+	for_each_bit(opt, i) {
+		if (bit_isset(opt, i)) {
+			flags[num_flags] = 1 << i;
+			(*combos)[num_comb++] = flags[num_flags++];
+		}	
+	}
+
+	start = 0;
+	end = num_flags;	
+	for (i = 0; i < num_flags; i++) {
+		for (j = start; j < end; j++) {
+			for (k = 0; k < num_flags; k++) {
+				toAdd = (*combos)[j] | flags[k];
+				if (!contains_pattern(*combos, num_comb, toAdd))
+					(*combos)[num_comb++] = toAdd;
+			}	
+		}	
+		start = end;
+		end = num_comb - 1;
+	}
```

Comment:
Please see my previous comment -- I believe this can be simplified to 2 nested loops.  The outer loop iterates over the number of combinations.  Each value represents the corresponding flags that should be set.  I.e. 0 - no flags, 1 - flag[0], 2 - flag[1], 3 - flag[0] | flag[1].